src/renderers/webgl/WebGLPrograms.d.ts

```diff
@@ -2,28 +2,158 @@ import { WebGLRenderer } from './../WebGLRenderer';
 import { WebGLProgram } from './WebGLProgram';
 import { WebGLCapabilities } from './WebGLCapabilities';
 import { WebGLExtensions } from './WebGLExtensions';
+import {
+	TextureEncoding,
+	Mapping,
+	Combine,
+	ToneMapping,
+	ShadowMapType,
+} from './../../constants';
 import { Material } from './../../materials/Material';
 import { Scene } from './../../scenes/Scene';
 
+export interface WebGLProgramParameters {
```

Comment:
Looking at `WebGLProgramParameters`, I'm don't think we want to expose all these properties in `onBeforeCompile()`. They are not intended for public access. The only relevant properties in context of `onBeforeCompile()` are `uniforms`, `vertexShader` and `fragmentShader`.

Reply:
Totally agree, but this is what is currently exposed. I just added the typings for it.
The difference for the user would be that compilation would break if `WebGLProgramParameters` changes in the future whereas without typings the code would break at runtime what is much worse.
I don't know what are the the API policies in this project. Are properties that are not part of typings treated as internal?

If you prefer I could also add `@internal`tag to the properties that should not be used. See
https://api-extractor.com/pages/tsdoc/tag_internal/

Reply:
Couldn't we just use the existing `Shader` definition?

https://github.com/mrdoob/three.js/blob/612f544c0c42bb01244472ffd57d0c88a37275b3/src/materials/Material.d.ts#L328

BTW: It seems your PR is not based on the latest commit in `dev`. I suggest you rebase.

Reply:
I think we'd prefer not to maintain type definitions for internal APIs. JavaScript exposes lots of things — including "private" methods on classes compiled from TypeScript — that don't need to provide type hints.

Let's type only the properties meant for public access, and add a comment above the interface itself explaining that the rest are internal.

Reply:
^Oops, I commented while @Mugen87 was replying. That suggestion looks good to me, too.

Reply:
@donmccurdy So the answer to 
> Are properties that are not part of typings treated as internal?

is a clear yes then.

@Mugen87 Regarding re-using the `Shader` type: `Shader` has `uniforms` whereas `WebGLProgramParameters` only has `defines` so it would not be correct to use `Shader` for `onBeforeCompile`. I will strip down `WebGLProgramParameters` to only contain `fragmentShader` and `vertexShader` for now. If we later think more stuff should be exposed we can add it.

Reply:
It should contain `uniforms`. Check out how the official example makes use of it:

https://threejs.org/examples/webgl_materials_modified

Reply:
@Mugen87  You are right. Before the parameters are send to `onBeforeCompile` the uniforms are added to the `parameters`.
So basically the type passed to `onBeforeCompile` and the types used in `WebGLPrograms.d.ts` are different. So either we abandon this PR or we change `Shader` to inherit from `WebGLProgramParameters` but I think it's not worth it b/c `WebGLProgramParameters` is internal.

Reply:
I don't think the typings for `onBeforeCompile()` require a change. However, improving `WebGLPrograms.d.ts` would be still nice.

> WebGLProgramParameters is internal.

Yes.